### PR TITLE
Add type hints

### PR DESCRIPTION
**Pull request by LarsV123**
Adds type hints to function parameters where the type could be easily inferred. Adding these should make further development easier.

Uses `from typing import List` to type collections, as a workaround to support Python versions below 3.9.

### CODE REVIEW

File: itkufs/accounting/management/commands/sendmail.py [hunk 1]
```diff
@@ -143,7 +146,7 @@ def _send_blocked_mail(self, account):
             subject, message, FROM_EMAIL, [to_email], headers=headers
         )
 
-    def _print_debug(self, emails):
+    def _print_debug(self, emails: list[str]):
```
Comment (arnels) [inline itkufs/accounting/management/commands/sendmail.py:149]:
```suggestion
    def _print_debug(self, emails: ListType[str]):
```

File: itkufs/accounting/views/edit.py [hunk 2]
```diff
@@ -28,7 +35,12 @@
 
 @login_required
 @limit_to_admin
-def new_edit_settlement(request, group, settlement=None, is_admin=False):
+def new_edit_settlement(
+    request: HttpRequest,
+    group: Group,
+    settlement: Settlement = None,
```
Comment (arnels) [inline itkufs/accounting/views/edit.py:41]:
```suggestion
    settlement: Optional[Settlement] = None,
```

File: itkufs/accounting/views/edit.py [hunk 3]
```diff
@@ -73,10 +85,10 @@ def new_edit_settlement(request, group, settlement=None, is_admin=False):
...
 @db_transaction.atomic
 def transfer(
-    request,
-    group,
-    account=None,
-    transfer_type=None,
+    request: HttpRequest,
+    group: Group,
+    account: Transaction = None,
+    transfer_type: str = None,
```
Comment (arnels) [inline itkufs/accounting/views/edit.py:91]:
```suggestion
    transfer_type: Optional[str] = None,
```

File: itkufs/accounting/views/edit.py [hunk 4]
```diff
@@ -73,10 +85,10 @@ def new_edit_settlement(request, group, settlement=None, is_admin=False):
 @limit_to_owner
 @db_transaction.atomic
 def transfer(
-    request,
-    group,
-    account=None,
-    transfer_type=None,
+    request: HttpRequest,
+    group: Group,
+    account: Transaction = None,
```
Comment (arnels) [inline itkufs/accounting/views/edit.py:90]:
```suggestion
    account: Optional[Account] = None,
```

File: itkufs/accounting/views/edit.py [hunk 5]
```diff
@@ -248,7 +262,12 @@ def approve_transactions(request, group, page="1", is_admin=False):
 
 @login_required
 @limit_to_owner
-def reject_transactions(request, group, transaction=None, is_admin=False):
+def reject_transactions(
+    request: HttpRequest,
+    group: Group,
+    transaction: Transaction = None,
```
Comment (arnels) [inline itkufs/accounting/views/edit.py:268]:
```suggestion
    transaction: Optional[Transaction] = None,
```

File: itkufs/accounting/views/edit.py [hunk 6]
```diff
@@ -325,7 +344,12 @@ def reject_transactions(request, group, transaction=None, is_admin=False):
 @login_required
 @limit_to_admin
 @db_transaction.atomic
-def new_edit_transaction(request, group, transaction=None, is_admin=False):
+def new_edit_transaction(
+    request: HttpRequest,
+    group: Group,
+    transaction: Transaction = None,
```
Comment (arnels) [inline itkufs/accounting/views/edit.py:350]:
```suggestion
    transaction: Optional[Transaction] = None,
```

File: itkufs/billing/views.py [hunk 7]
```diff
@@ -19,7 +19,9 @@
 
 @login_required
 @limit_to_admin
-def bill_new_edit(request, group, bill=None, is_admin=False):
+def bill_new_edit(
+    request: HttpRequest, group: Group, bill: Bill = None, is_admin=False
```
Comment (arnels) [inline itkufs/billing/views.py:23]:
```suggestion
    request: HttpRequest, group: Group, bill: Optional[Bill] = None, is_admin=False
```

File: itkufs/common/views/edit.py [hunk 8]
```diff
@@ -58,7 +58,11 @@ def edit_group(request, group, is_admin=False):
 
 @login_required
 def activate_account(
-    request, group, account=None, is_admin=False, is_owner=False
+    request: HttpRequest,
+    group: Group,
+    account: Account = None,
```
Comment (arnels) [inline itkufs/common/views/edit.py:63]:
```suggestion
    account: Optional[Account] = None,
```

File: itkufs/common/views/edit.py [hunk 9]
```diff
@@ -77,7 +81,11 @@ def activate_account(
 @login_required
 @limit_to_admin
 def new_edit_account(
-    request, group, account=None, is_admin=False, is_owner=False
+    request: HttpRequest,
+    group: Group,
+    account: Account = None,
```
Comment (arnels) [inline itkufs/common/views/edit.py:86]:
```suggestion
    account: Optional[Account] = None,
```

File: itkufs/reports/forms.py [hunk 10]
```diff
@@ -119,7 +119,7 @@ def clean(self):
             )
         return self.cleaned_data
 
-    def save(self, group=None, **kwargs):
+    def save(self, group: Group = None, **kwargs):
```
Comment (arnels) [inline itkufs/reports/forms.py:122]:
```suggestion
    def save(self, group: Optional[Group] = None, **kwargs):
```

**Comment by arnels** [comment]
Great initiative! We should probably enable mypy in tox.ini

File: itkufs/accounting/views/edit.py [hunk 11]
```diff
@@ -73,10 +87,10 @@ def new_edit_settlement(request, group, settlement=None, is_admin=False):
 @limit_to_owner
 @db_transaction.atomic
 def transfer(
-    request,
-    group,
-    account=None,
-    transfer_type=None,
+    request: HttpRequest,
+    group: Group,
+    account: Optional[Account] = None,
```
Comment (LarsV123) [inline itkufs/accounting/views/edit.py:92]: Can `account` ever be `None` here @arnels ? If so, what is an appropriate error to raise?

**Comment by LarsV123** [comment]
> Great initiative! We should probably enable mypy in tox.ini

I added `mypy` (but not to `tox.ini`) to try it. It mainly complains about:
1. Missing types from imported libraries (ignored for now).
2. Possible `None` values for various view methods, being used without checking if they are `None`. Perhaps throw an error here or remove the `Optional` from them if they can never be `None`?
3. No type hinting for the property system in `itkufs/accounting/models.py`.

Any suggestions for how to fix these?

**Review by arnels — APPROVED** [review]

File: itkufs/accounting/views/edit.py (hunk 11 above)
Comment (LarsV123) [reply itkufs/accounting/views/edit.py]: I changed this and a few other things mypy complained about. I'm not able to resolve the remaining mypy complaints and suspect it requires a newer Django version to do so, so you can merge this now if you want.

**Pull request merged**